ompi/runtime/ompi_spc.c

```diff
@@ -0,0 +1,510 @@
+        *count = 1;
+    }
+    /* For this event, we need to turn on the counter */
+    else if(MCA_BASE_PVAR_HANDLE_START == event) {
+        /* Convert from MPI_T pvar index to SPC index */
+        index = pvar->pvar_index - mpi_t_offset;
+        attached_event[index] = 1;
+    }
+    /* For this event, we need to turn off the counter */
+    else if(MCA_BASE_PVAR_HANDLE_STOP == event) {
+        /* Convert from MPI_T pvar index to SPC index */
+        index = pvar->pvar_index - mpi_t_offset;
+        attached_event[index] = 0;
+    }
+
+    return MPI_SUCCESS;
+}
+
+/* ##############################################################
+ * ################# Begin SPC Functions ########################
+ * ##############################################################
+ */
+
+/* This function returns the current count of an SPC counter that has been retistered
+ * as an MPI_T pvar.  The MPI_T index is not necessarily the same as the SPC index,
+ * so we need to convert from MPI_T index to SPC index and then set the 'value' argument
+ * to the correct value for this pvar.
+ */
+static int ompi_spc_get_count(const struct mca_base_pvar_t *pvar, void *value, void *obj_handle)
+{   
+    (void) obj_handle;
+
+    long long *counter_value = (long long*)value;
+
+    if(OPAL_UNLIKELY(mpi_t_disabled == 1)) {
+        *counter_value = 0;
+        return MPI_SUCCESS;
+    }
+
+    /* Convert from MPI_T pvar index to SPC index */
+    int index = pvar->pvar_index - mpi_t_offset;
+    /* Set the counter value to the current SPC value */
+    *counter_value = events[index].value;
+    /* If this is a timer-based counter, convert from cycles to microseconds */
+    if(timer_event[index])
+        *counter_value /= sys_clock_freq_mhz;
+    /* If this is a high watermark counter, reset it after it has been read */
+    if(index == OMPI_MAX_UNEXPECTED_IN_QUEUE || index == OMPI_MAX_OOS_IN_QUEUE)
+        events[index].value = 0;
+
+    return MPI_SUCCESS;
+}
+
+/* Initializes the events data structure and allocates memory for it if needed. */
+void events_init()
+{
+    int i;
+
+    /* If the events data structure hasn't been allocated yet, allocate memory for it */
+    if(events == NULL) {
+        events = (ompi_event_t*)malloc(OMPI_NUM_COUNTERS * sizeof(ompi_event_t));
+    }
+    /* The data structure has been allocated, so we simply initialize all of the counters
+     * with their names and an initial count of 0.
+     */
+    for(i = 0; i < OMPI_NUM_COUNTERS; i++) {
+        events[i].name = counter_names[i];
+        events[i].value = 0;
+    }
+}
+
+/* Initializes the SPC data structures and registers all counters as MPI_T pvars.
+ * Turns on only the counters that were specified in the mpi_spc_attach MCA parameter.  
+ */
+void ompi_spc_init()
+{
+    int i, j, ret, found = 0, all_on = 0;
+
+    /* Initialize the clock frequency variable as the CPU's frequency in MHz */
+    sys_clock_freq_mhz = opal_timer_base_get_freq() / 1000000;
```

Comment:
What happens if the system clock frequency is variable?

Should we just be using the high-resolution clock, instead?

Reply:
These timers are short lived, they basically measure the cost of a match. So we prefer something fast to something slower but more accurate.

Reply:
I recently had a very similar discussion with a system administrator, but on the subject of continuous integration testing on github.

In general, the CI tests were great (just like they are here on the OMPI project): you can get cheap/free testing across a wide variety of platforms just by pushing a git commit.  The developers came to rely on these tests to exercise platforms that they did not have easy access to.

Over the course of a week, dozens of PRs were merged.

Suddenly, QA reported back that the software was horribly broken on several platforms.

When the developers looked into what happened, they were very confused: all the problems *should* have been caught by CI.

It turns out that the system administrators were focused on delivering *extremely fast* continuous integration tests.  Meaning: if a test was going to take too long, it was killed, and the test simply reported success.  This led to a lot of tests not actually being run, and therefore a lot of bugs slipped through.

In short: developers were relying on behavior that was not actually happening.  They got their CI results *fast*, but they weren't *accurate*.

A similar situation exists here: there are many reasons that the CPU frequency can change -- even a little bit -- and therefore these counters won't actually reflect reality.  But in the reported results, there will be no way to tell that.

What use is that?